calyx/frontend/src/parser.rs

```diff
@@ -869,24 +878,133 @@ impl CalyxParser {
         ))
     }
 
-    fn connections(
-        input: Node,
-    ) -> ParseResult<(Vec<ast::Wire>, Vec<ast::Group>, Vec<ast::StaticGroup>)>
-    {
+    /// Parses in the state indices, which are unsigned integers in range `0..n`
+    fn state_idx(input: Node) -> ParseResult<u64> {
+        input
+            .as_str()
+            .parse::<u64>()
+            .map_err(|_| input.error("Expected valid state index"))
+    }
+
+    /// Parses the conditional transitions, which are of form `GuardExpr ->` state index,
+    /// or the default case, `default ->` state index. The conditional transitions
+    /// are evaluated sequentially, so the default case is treated like an `else` or
+    /// like the default case of a verilog case statement.
+    fn guard_state_pair(input: Node) -> ParseResult<(ast::Guard, u64)> {
+        let span = Self::get_span(&input);
+        Ok(match_nodes!(
+            input.into_children();
+            [guard_expr(guard_expr), state_idx(state)] => { // guard branch
+                let guard = ast::Guard {
+                    guard: Some(*guard_expr),
+                    expr: ast::Atom::Num(BitNum {
+                        width: 1,
+                        num_type: NumType::Decimal,
+                        val: 1,
+                        span,
+                    }),
```

Comment:
It looks like the `expr` part of the `ast::Guard` struct is unused here, in all cases. That is, we are just filling this `expr` part in with `1'd1` and never actually using that, and the only interesting stuff is in the `guard` field.

Somewhat confusingly, it appears that the `Guard` struct is inaccurately named: it should be called `GuardedExpr` or similar. What we actually want to return here is just the guard itself, i.e., an `ast::GuardExpr` or possibly an `Option<ast::GuardExpr>`, because that is the only meaningful thing we're parsing here.

Can you please change the return type of this function to be `GuardExpr` or `Option<GuardExpr>`, and change the `Transition` struct accordingly to use the same?